doc/sonic-application-extension/tpcm_app_ext.md

```diff
@@ -0,0 +1,467 @@
+- Ensure TPC has right system resources and privileges
+- Abilitiy to configure startup command arguments to TPC
+- Configure resource limits to TPCs, to ensure it does not starve core SONiC containers
+- Ability to seamlessly integrate into SONiC Services Architecture enabling start/ auto restart/ establish dependencies
+
+
+### Requirements
+
+These open TPCs help extend SONiC capabilities, and thus the following requirements are outlined to integrate them seamlessly into SONiC
+
+- Dynamically install TPC from docker registry and docker image from the local file system, SCP, SFTP, URL. These remote download options are introduced now to SONiC AEF.
+- Upgrade TPCs from docker registry and docker image from local file system, SCP, SFTP, URL
+- Provide default manifest file for docker image without manifest during install.
+- Provide runtime install capability to pass various docker startup arguments and parameters through local custom manifest file.
+- Specify system resource limits for TPCs to restrict CPU, Memory usage
+- Provide update capability to update various TPC configurations like their memory, cpu, dependencies etc. This capability is new to SONiC AEF.
+
+
+
+### Architecture Design 
+
+There is no change to the SONiC architecture. 
+
+
+### High-Level Design 
+
+SONiC Application Extension Infrastructure provides the framework to integrate SONiC compatible dockers. However there are many open source third party applications which can be installed on SONiC system, to extend the capabilities, and these typically are standalone or have less interaction with the SONiC system itself. So it is not necesary to pre define these docker applications in SONiC build, and provide their corresponding manifest.json. These TPCs can be installed dynamically on a SONiC device and can be managed.
+
+- This feature enables the installation and management of third-party Docker packages without manifest files through the Sonic Package Manager. 
+- Also, it manages the TPC without affecting the sonic system resources.
+- This feature enables users to create a local custom manifest file (from default manifest template file) on-the-fly which can be updated or deleted as needed. 
+- Additionally, the Sonic Package Manager provides CLI commands to create, update, and delete the local manifest file.
+
+
+<!-- omit in toc -->
+###### Figure. SONiC TPCM Support
+
+<p align=center>
+<img src="img/TPCM_Flow_Diagram.png" alt="Figure. SONiC TPCM Flow Diagram">
+</p>
+
+
+
+#### TPC Install
+
+- The TPCM support feature will leverage the existing Sonic Package Manager framework.
+- Ability to download image tarballs for the Docker packages through SCP, SFTP, and URL before installing them is introduced.
+- If a manifest file is not found in the docker image during installation, a default local TPC manifest file is used to install the package.
+- A Docker image that does not contain a manifest file will be considered a Third-Party Container (TPC).
```

Comment:
IMO, we should not differentiate like this, wherever possible we should design generically, in this case, App ext can be extended to support docker image without manifest file as well.

Reply:
Agreed